# Exercise 4.5.

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [2]:
path=r'C:\Users\User\Achievement4\Data'

In [3]:
# Import "products" file
df_prods = pd.read_csv(os.path.join(path, 'Original Data', 'products.csv'))

In [4]:
# Import "orders" file
df_ords = pd.read_csv(os.path.join(path, 'Prepared Data', 'orders_wrangled.csv'))

In [5]:
# Create a dataframe
df_test=pd.DataFrame()

In [6]:
# Create a mix type column
df_test['mix']=['a', 'b', 1, True]


In [7]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [8]:
#Check for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


C:\Users\User\AppData\Local\Temp\ipykernel_6932\522087711.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)


In [9]:
df_test['mix'] = df_test['mix'].astype('str')

In [10]:
# Checking for missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [11]:
# Creating subset with missing values
ds_nan=df_prods[df_prods['product_name'].isnull()==True]

In [12]:
ds_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [13]:
df_prods.shape

(49693, 5)

In [14]:
# Creating new data frame without missing values
df_prods_clean=df_prods[df_prods['product_name'].isnull()==False]

In [15]:
df_prods_clean.shape

(49677, 5)

In [16]:
# Creating a subset with duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [17]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [18]:
df_dups.shape

(5, 5)

In [19]:
df_prods_clean.shape

(49677, 5)

In [20]:
# Creating a new subset without duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [21]:
df_prods_clean_no_dups.shape

(49672, 5)

In [22]:
# Export df_prods_clean_no_dups dataframe
df_prods_clean_no_dups.to_csv(os.path.join(path,'Prepared Data','products_clean.csv'))

# Task

In [37]:
df_prods_clean_no_dups.describe()

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.349775     67.762442      11.728942      9.993282
std    14340.705287     38.315784       5.850779    453.615536
min        1.000000      1.000000       1.000000      1.000000
25%    12432.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

## 1)'orders_day_of_week' has a minimum of '0' and a maximum of '6' which means the days are zero-indexed. 2) 'days_since_prior_order' has a maximum value of 30. I assume that's because the orders are capped monthly. 3) The lowest value for 'days_since_prior_order' is 0, indicating either multiple orders placed on the same day or serving as a placeholder for a customer's first order."

In [24]:
# 3.Checking for mix type data
for col in df_ords.columns.tolist():
    weird = (df_ords[col].map(type) != df_ords[col].iloc[0].__class__).any()
    if weird:
        print(col)


Unnamed: 0
order_id
user_id
order_number
orders_day_of_week
hour_of_day
days_since_prior_order


In [25]:
df_ords.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 7 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   Unnamed: 0              int64  
 1   order_id                int64  
 2   user_id                 int64  
 3   order_number            int64  
 4   orders_day_of_week      int64  
 5   hour_of_day             int64  
 6   days_since_prior_order  float64
dtypes: float64(1), int64(6)
memory usage: 182.7 MB


## There are no columns with mix type data

In [26]:
# 5.Checking for missing values
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
hour_of_day                    0
days_since_prior_order    206209
dtype: int64

## As we see, there are 206209 cells in "days_since_prior_order" with missing values

In [27]:
# 6. Creating a subset containing only missing values
df_ords_nan=df_ords[df_ords['days_since_prior_order'].isnull()==True]

In [28]:
df_ords_nan

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
11               11   2168274        2             1                   2   
26               26   1374495        3             1                   1   
39               39   3343014        4             1                   6   
45               45   2717275        5             1                   3   
...             ...       ...      ...           ...                 ...   
3420930     3420930    969311   206205             1                   4   
3420934     3420934   3189322   206206             1                   3   
3421002     3421002   2166133   206207             1                   6   
3421019     3421019   2227043   206208             1                   1   
3421069     3421069   3154581   206209             1                   3   

         hour_of_day  days_since_prior_order  
0                  8                     NaN  
11                11                     NaN  
26                14                     NaN  
39                11                     NaN  
45                12                     NaN  
...              ...                     ...  
3420930           12                     NaN  
3420934           18                     NaN  
3421002           19                     NaN  
3421019           15                     NaN  
3421069           11                     NaN  

[206209 rows x 7 columns]

## These are 206209 missing values in the 'days_since_prior_order" column. Let's check if all rows with missing values in 'days_since_prior_order" column have "1" in "order_number"

In [29]:
df_ords_nan['order_number'].value_counts()

order_number
1    206209
Name: count, dtype: int64

## All rows in the df_ords_nan subset contain "1" in "order_number", which means all of them are the first orders

## Checking how many first orders there are in the main dataframe df_ords

In [30]:
df_ords['order_number'].value_counts()

order_number
1      206209
2      206209
3      206209
4      206209
5      182223
        ...  
96       1592
97       1525
98       1471
99       1421
100      1374
Name: count, Length: 100, dtype: int64

## In df_ords there are 206209 orders with "1" value, now  we are 100% sure that the frequency of missing values (206,209) is the same as the frequency of order_number=1.

In [31]:
# Address the missing values by creating a column that flags order numbers as either True/False depending on if the 'order_number' = 1
df_ords['first_order'] = df_ords['days_since_prior_order'].isna()

In [32]:
df_ords

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
1                 1   2398795        1             2                   3   
2                 2    473747        1             3                   3   
3                 3   2254736        1             4                   4   
4                 4    431534        1             5                   4   
...             ...       ...      ...           ...                 ...   
3421078     3421078   2266710   206209            10                   5   
3421079     3421079   1854736   206209            11                   4   
3421080     3421080    626363   206209            12                   1   
3421081     3421081   2977660   206209            13                   1   
3421082     3421082    272231   206209            14                   6   

         hour_of_day  days_since_prior_order  first_order  
0                  8                     NaN         True  
1                  7                    15.0        False  
2                 12                    21.0        False  
3                  7                    29.0        False  
4                 15                    28.0        False  
...              ...                     ...          ...  
3421078           18                    29.0        False  
3421079           10                    30.0        False  
3421080           12                    18.0        False  
3421081           12                     7.0        False  
3421082           14                    30.0        False  

[3421083 rows x 8 columns]

In [33]:
# 7.Looking for duplicates
df_dups_ords = df_ords[df_ords.duplicated()]

In [34]:
df_dups_ords

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, hour_of_day, days_since_prior_order, first_order]
Index: []

## There are no duplicates in df_ords

In [35]:
# 9. Exporting dataframe df_ords with clean data (The dataframe df_prods has already been exported during the Exercise)
df_ords.to_csv(os.path.join(path,'Prepared Data','orders_cleaned.csv'))